docs/flip/1009-optimizer-api.md

```diff
@@ -0,0 +1,529 @@
+we define the `Optimizer` as a named tuple (instead of a
+`flax.struct.dataclass`):
+
+```python
+# flax/opt/base.py
+
+Params = Any
+Gradients = Any
+OptimizerState = NamedTuple
+
+
+class Optimizer(NamedTuple):
+  init: Callable[[Params], OptimizerState]
+  update: Callable[
+      [Gradients, Params, OptimizerState], Tuple[Params, OptimizerState]]
+```
+
+For example, to implement `flax.optim.Momentum` in this new interface:
+
+```python
+# flax/opt/momentum.py
+
+class MomentumState(OptimizerState):
+  momentum: PyTree
+
+
+def momentum(learning_rate: float, beta: float = 0.9) -> Optimizer:
+
+  def init_fn(params: Params) -> OptimizerState:
+    return MomentumState(momentum=jax.tree_map(jnp.zeros_like, params))
+
+  def update_fn(
+    grads: Gradients, params: Params, state: OptimizerState
+  ) -> Tuple[Params, OptimizerState]:
+    new_momentum = jax.tree_multimap(
+        lambda momentum, grad: momentum * beta + grad,
+        opt_state.momentum, grads)
+    new_params = jax.tree_multimap(
+        lambda param, grad: param - learning_rate * grad,
+        params, grads)
+    return new_params, MomentumState(new_momentum)
+
+  return Optimizer(init_fn, update_fn)
+```
+
+Note that unlike in the [previous API], the function `update_fn()` will be
+called directly in the train loop and must take care of processing the pytrees
+of parameters and optimizer state with a `jax.tree_multimap()`. For more
+complex optimizers this can become a hassle but we can provide a small
+functional helper that takes care of this:
+
+```python
+  def update_fn(
+    grads: Gradients, params: Params, state: OptimizerState
+  ) -> Tuple[Params, OptimizerState]:
+
+    @multimap
+    def inner(param, grad, momentum):
+      new_momentum = momentum * beta + grad
+      new_param = param - learning_rate * grad
+      return new_param, new_momentum
+
+    new_params, new_momentum = inner(params, grads, state.momentum)
+    return new_params, MomentumState(new_momentum)
+```
+
+Remarks:
+
+- We'll call `Optimizer` instances from the functional API `optim`, and refer
+  to `flax.optim.Optimizer` instances from the [previous API] as `optimizer`
+  for the remainder of this document.
+- As opposed to the [previous API], this API does not have the option to update
+  hyper parameters. When [using Optax], updating the hyper parameters is instead
+  done by defining a schedule that takes care of updating the hyper parameters
+  during training.
+- Using this simple API directly makes the user code a bit more verbose than
+  using the [previous API], but the difference is smaller with Linen. If you
+  care about a boilerplate-free training loop see the proposed [Linen helper]
+  further down.
+
+## Using Optax
+[Using Optax]: #using-optax
+
+Optax is centered on the idea of composable gradient transformations. Since
+above [functional API] is based on Optax's interface it is straightforward to
+wrap an Optax transformation to comply with the `Optimizer` interface:
```

Comment:
Suggestion: nit

```suggestion
Optax is centered on the idea of composable gradient transformations. Since
the above-mentioned [functional API] is based on Optax's interface it is straightforward to
wrap an Optax transformation to comply with the `Optimizer` interface:
```

("above-mentioned" or the "new"/"proposed")

Reply:
Updated.